drivers/Aqara/aqara-lock/src/credential_utils.lua

```diff
@@ -0,0 +1,151 @@
+  for _, credentialinfo in ipairs(command.args.credentialInfo) do
+    if credentialinfo.userType == "host" then
+      update_remote_control_status(driver, device, true)
+    end
+  end
+  device:emit_event(lockCredentialInfo.credentialInfo(utils.deep_copy(command.args.credentialInfo), { visibility = { displayed = false } }))
+  save_data(driver)
+end
+
+local function upsert_credential_info(driver, device, command)
+  if #command.args.credentialInfo == 0 then
+    return
+  end
+
+  local credentialInfoTable = utils.deep_copy(device:get_latest_state("main", lockCredentialInfo.ID, lockCredentialInfo.credentialInfo.NAME, {}))
+
+  for _, credentialinfo in ipairs(command.args.credentialInfo) do
+    local exist = false
+    for i = 1, #credentialInfoTable, 1 do
+      if credentialInfoTable[i].credentialId == credentialinfo.credentialId then
+        credentialInfoTable[i] = utils.deep_copy(credentialinfo)
+        exist = true
+        break
+      end
+    end
+
+    if exist == false then
+      if credentialinfo.userType == "host" then
+        update_remote_control_status(driver, device, true)
+      end
+
+      table.insert(credentialInfoTable, credentialinfo)
+    end
+  end
+
+  device:emit_event(lockCredentialInfo.credentialInfo(utils.deep_copy(credentialInfoTable), { visibility = { displayed = false } }))
+  save_data(driver)
+end
+
+local function delete_user(driver, device, command)
+  local credentialInfoTable = utils.deep_copy(device:get_latest_state("main", lockCredentialInfo.ID, lockCredentialInfo.credentialInfo.NAME, {}))
+  if #credentialInfoTable == 0 then
+    return
+  end
+
+  local id = tostring(command.args.userId)  
+
+  for i = #credentialInfoTable, 1, -1 do
+    if id == credentialInfoTable[i].userId then
+      if credentialInfoTable[i].userType == "host" then
+        update_remote_control_status(driver, device, false)
+      end
+      table.remove(credentialInfoTable, i)
+    end
+  end
+
+  device:emit_event(lockCredentialInfo.credentialInfo(utils.deep_copy(credentialInfoTable), { visibility = { displayed = false } }))
+  save_data(driver)
+end
+
+local function delete_credential(driver, device, command)
+  local credentialInfoTable = utils.deep_copy(device:get_latest_state("main", lockCredentialInfo.ID, lockCredentialInfo.credentialInfo.NAME, {}))
+  if #credentialInfoTable == 0 then
+    return
+  end
+
+  for i = #credentialInfoTable, 1, -1 do
+    if command.args.credentialId == credentialInfoTable[i].credentialId then
+      if credentialInfoTable[i].userType == "host" then
+        update_remote_control_status(driver, device, false)
+      end
+      table.remove(credentialInfoTable, i)
+      break
+    end
+  end
+
+  device:emit_event(lockCredentialInfo.credentialInfo(utils.deep_copy(credentialInfoTable), { visibility = { displayed = false } }))
+  save_data(driver)
+end
+
+local function update_system_version(driver, device, command)
+  local version = command.args.version
+end
```

Comment:
this function is a no-op

Reply:
@seojune79 Could you please remove this function? this wouldn't be used anymore.

Reply:
I removed it.